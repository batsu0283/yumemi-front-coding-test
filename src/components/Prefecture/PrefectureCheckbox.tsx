import { memo } from "react";
import type { PrefCode } from "../../utils/constants";

type PrefectureCheckboxProps = {
  prefCode: PrefCode;
  prefName: string;
  checked: boolean;
  handlePrefectureCheckBox: (prefCode: PrefCode) => void;
};

export const PrefectureCheckbox = memo(
  ({
    prefCode,
    prefName,
    checked,
    handlePrefectureCheckBox,
  }: PrefectureCheckboxProps) => (
    <label className="flex items-center gap-2 p-2 bg-gray-100 rounded-lg shadow-sm">
      <input
        type="checkbox"
        className="w-5 h-5 text-cyan-600 focus:ring-cyan-500"
        checked={checked}
        onChange={() => handlePrefectureCheckBox(prefCode)}
        aria-label={`都道府県 ${prefName} を選択`}
      />
      <span className="text-gray-700">{prefName}</span>
    </label>
  ),
);
