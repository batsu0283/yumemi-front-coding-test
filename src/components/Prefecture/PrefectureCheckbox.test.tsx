import { fireEvent, render, screen } from "@testing-library/react";
import type { PrefCode } from "../../utils/constants";
import { PrefectureCheckbox } from "./PrefectureCheckbox";

describe("PrefectureCheckbox", () => {
  const mockHandleCheckboxChange = vi.fn();
  const prefCode: PrefCode = 1;
  const prefName = "北海道";

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("都道府県名が正しく表示されること", () => {
    render(
      <PrefectureCheckbox
        prefCode={prefCode}
        prefName={prefName}
        checked={false}
        handleCheckboxChange={mockHandleCheckboxChange}
      />,
    );
    expect(screen.getByText(prefName)).toBeInTheDocument();
    const checkbox = screen.getByRole("checkbox", {
      name: `都道府県 ${prefName} を選択`,
    });
    expect(checkbox).not.toBeChecked();
  });

  it("チェック状態が反映されること", () => {
    render(
      <PrefectureCheckbox
        prefCode={prefCode}
        prefName={prefName}
        checked={true}
        handleCheckboxChange={mockHandleCheckboxChange}
      />,
    );
    const checkbox = screen.getByRole("checkbox", {
      name: `都道府県 ${prefName} を選択`,
    });
    expect(checkbox).toBeChecked();
  });

  it("チェックボックスの状態が変わった時に handleCheckboxChange が呼ばれること", () => {
    render(
      <PrefectureCheckbox
        prefCode={prefCode}
        prefName={prefName}
        checked={false}
        handleCheckboxChange={mockHandleCheckboxChange}
      />,
    );
    const checkbox = screen.getByRole("checkbox", {
      name: `都道府県 ${prefName} を選択`,
    });

    fireEvent.click(checkbox);
    expect(mockHandleCheckboxChange).toHaveBeenCalledWith(prefCode);
    expect(mockHandleCheckboxChange).toHaveBeenCalledTimes(1);
  });

  it("Snapshot", () => {
    const { container } = render(
      <PrefectureCheckbox
        prefCode={prefCode}
        prefName={prefName}
        checked={false}
        handleCheckboxChange={mockHandleCheckboxChange}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
