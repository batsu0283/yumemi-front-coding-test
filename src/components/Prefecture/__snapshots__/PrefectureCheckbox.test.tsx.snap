// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrefectureCheckbox > Snapshot 1`] = `
<div>
  <label
    class="flex items-center gap-2 p-2 bg-gray-100 rounded-lg shadow-sm"
  >
    <input
      aria-label="都道府県 北海道 を選択"
      class="w-5 h-5 text-cyan-600 focus:ring-cyan-500"
      type="checkbox"
    />
    <span
      class="text-gray-700"
    >
      北海道
    </span>
  </label>
</div>
`;
